docs/design/sql/04-parallel-execution.md

```diff
@@ -0,0 +1,146 @@
+`threadIndex = partition % threadCount`.
+
+The partition pool has the following advantages:
+1. Only one thread processes messages with the given partition so that processing logic may use less synchronization.
+1. Dedicated thread queues reduce contention on enqueue/deque operations.
+
+However, there is no load balancing in the partition pool: a single long-running task may delay other tasks from the same 
+partition indefinitely. An imbalance between partitions may cause low resource utilization.
+
+The partition pool is thus suitable for small tasks that operate on independent physical resources, and that are
+distributed equally between logical partitions. An example is `IMap` operations, which operate on separate physical
+partitions, such as `GET` and `PUT`.
+
+Since the partition is a logical notion, it is possible to multiplex tasks from different components to a single partition pool.
+For example, CP Subsystem schedules tasks, all with the same partition, to the partition pool to ensure total processing order.
+
+### 1.3 Generic Pool
+
+If a message doesn't have a logical partition, it is submitted to the **generic pool**. This is a conventional thread pool with
+a shared blocking queue. It has inherent balancing capabilities. At the same time, this pool may demonstrate less than
+optimal throughput when a lot of small tasks are submitted due to contention on the queue.
+
+### 1.4 Hazelcast Jet Pool
+
+Hazelcast Jet uses its own cooperative pool to execute Jet jobs. Every thread has its own queue of jobs that are executed
+cooperatively. There is no balancing: once the job is submitted to a specific thread, it is always executed in that thread.
+
+IO pool doesn't notify the Jet pool about new data batch ("push"). Instead, the message is just enqueued, and the Jet thread
+checks the queue periodically ("poll").
+
+## 2 Design
+
+We now define the requirements to Hazelcast Mustang threading model, analyze them concerning existing infrastructure, and
+define the design.
+
+### 2.1 Requirements
+
+The requirements are thread safety, load balancing, and ordered processing.
+
+First, the infrastructure must guarantee that operator execution is thread-safe. That is, the stateful operator should not be
+executed by multiple threads simultaneously. This simplifies operator implementations and makes them more performant.
+Hazelcast Jet follows this principle, as only one thread may execute a particular job. However, Hazelcast Jet pool doesn't
+satisfy the load balancing requirement discussed below.
+
+Second, the execution environment must support load balancing. Query execution may take a long time to complete. If several query
+fragments have been assigned to a single execution thread, it should be possible to reassign them to idle threads dynamically.
+Neither partition pool nor Hazelcast Jet pool designs are applicable to Hazelcast Mustang because they lack balancing
+capabilities.
```

Comment:
The IMDG SQL is different - it may allow for a long execution of a certain fragment. During this time, we do not want another task sitting in the queue of the same thread to wait. This is where the load balancing comes into play.